lib/hooks/userspace/hook-koa.js

```diff
@@ -0,0 +1,130 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+var cls = require('../../cls.js');
+var TraceLabels = require('../../trace-labels.js');
+var shimmer = require('shimmer');
+var semver = require('semver');
+var constants = require('../../constants.js');
+var agent;
+
+var SUPPORTED_VERSIONS = '1.x';
+
+function useWrap(use) {
+  return function useTrace() {
+    if (!this._google_trace_patched) {
+      this._google_trace_patched = true;
+      this.use(middleware);
+    }
+    return use.apply(this, arguments);
+  };
+}
+
+function *middleware(next) {
+  var namespace = cls.getNamespace();
+  if (!namespace) {
+    agent.logger.info('Koa: no namespace found, ignoring request');
+    return;
+  }
+  if (!agent.shouldTrace(this.req.url)) {
+    return;
+  }
+  var req = this.req;
+  var res = this.res;
+
+  namespace.bindEmitter(req);
+  namespace.bindEmitter(res);
+
+  var originalEnd = res.end;
+
+  namespace.run(function() {
+    var rootContext = startRootSpanForRequest(req);
+
+    // wrap end
+    res.end = function(chunk, encoding) {
+      res.end = originalEnd;
+      var returned = res.end(chunk, encoding);
+
+      endRootSpanForRequest(rootContext, req, res);
+      return returned;
+    };
+    namespace.bind(next);
+  });
+  yield next;
+}
+
+/**
+ * Creates and sets up a new root span for the given request.
+ * @param {Object} req The request being processed.
+ * @returns {!SpanData} The new initialized trace span data instance.
+ */
+function startRootSpanForRequest(req) {
+  var result = agent.parseContextFromHeader(
+    req.headers[constants.TRACE_CONTEXT_HEADER_NAME]) || {};
+
+  var traceId = result.traceId;
+  var parentSpanId = result.spanId;
+  var url = (req.headers['X-Forwarded-Proto'] || 'http') +
+    '://' + req.headers.host + req.url;
+
+  // we use the path part of the url as the span name and add the full
+  // url as a label
+  var rootContext = agent.createRootSpanData(req.url, traceId,
+    parentSpanId);
+  rootContext.addLabel(TraceLabels.HTTP_METHOD_LABEL_KEY, req.method);
+  rootContext.addLabel(TraceLabels.HTTP_URL_LABEL_KEY, url);
+  rootContext.addLabel(TraceLabels.HTTP_SOURCE_IP, req.connection.remoteAddress);
+  return rootContext;
+}
+
+
+/**
+ * Ends the root span for the given request.
+ * @param {!SpanData} rootContext The span to close out.
+ * @param {Object} req The request being processed.
+ * @param {Object} res The response being processed.
+ */
+function endRootSpanForRequest(rootContext, req, res) {
+  if (req.route && req.route.path) {
```

Comment:
They happen to be present in express (and in koa apparently). We can steer away from using undocumented things but the koa documentation doesn't appear to be fully accurate (or at least doesn't describe when request properties will be undefined.